console-src/src/enhancers/withOAuthClients.tsx

```diff
@@ -0,0 +1,190 @@
+import React, { Component } from "react";
+import { Client } from "@kontist/client";
+
+import prodConfig from "../config";
+import devConfig from "../config.dev";
+
+import Loading from "../pages/Loading";
+
+import {
+  OAuthClient,
+  CreateOAuthClientPayload,
+  UpdateOAuthClientPayload,
+  DeleteOAuthClientPayload
+} from "../types/oAuthClient";
+import {
+  fetchClientsQuery,
+  createClientMutation,
+  updateClientMutation,
+  deleteClientMutation
+} from "../queries";
+
+const { baseAPIUrl, clientId, redirectUri } =
+  process.env.NODE_ENV === "production" ? prodConfig : devConfig;
+
+const STATE_KEY = "state";
+const VERIFIER_KEY = "verifier";
+const CLIENT_SCOPE = "clients";
+const CODE_QUERY_PARAM = "code";
+
+sessionStorage.setItem(
+  STATE_KEY,
+  sessionStorage.getItem(STATE_KEY) || (Math.random() + "").substring(2)
+);
+sessionStorage.setItem(
+  VERIFIER_KEY,
+  sessionStorage.getItem(VERIFIER_KEY) || (Math.random() + "").substring(2)
+);
+
+const client = new Client({
+  baseUrl: baseAPIUrl,
+  clientId,
+  redirectUri,
+  scopes: [CLIENT_SCOPE],
+  state: sessionStorage.getItem(STATE_KEY) || "",
+  verifier: sessionStorage.getItem(VERIFIER_KEY) || ""
+});
+
+interface OAuthClientsContext {
+  isLoading: boolean;
+  oAuthClients: OAuthClient[];
+  createClient: (payload: CreateOAuthClientPayload) => Promise<Object> | void;
+  updateClient: (payload: UpdateOAuthClientPayload) => Promise<Object> | void;
+  deleteClient: (payload: DeleteOAuthClientPayload) => Promise<Object> | void;
+}
+
+interface State extends OAuthClientsContext {
+  isAuthenticated: boolean;
+}
+
+const { Provider, Consumer } = React.createContext<OAuthClientsContext>({
+  isLoading: false,
+  oAuthClients: [],
+  createClient: () => {},
+  updateClient: () => {},
+  deleteClient: () => {}
+});
+
+class OAuthClientsProvider extends Component<{}, State> {
+  fetchClients: () => Promise<any>;
+  createClient: (payload: CreateOAuthClientPayload) => Promise<any>;
+  updateClient: (payload: UpdateOAuthClientPayload) => Promise<any>;
+  deleteClient: (payload: DeleteOAuthClientPayload) => Promise<any>;
+
+  constructor(props: {}) {
+    super(props);
+
+    this.fetchClients = async () => {
```

Comment:
Why are this methods defined in constructor instead of just having them in class like:
```
constructor() {
  ...
}

fetchClients = async () => {
...
}
```

Reply:
this is the recommended way in React docs to use context when you want to pass methods to consumers: https://reactjs.org/docs/context.html#updating-context-from-a-nested-component

I'm not 100% sure that it is required to do it this way, and it looks a little odd, but I'd rather stick to the official recommendation, what do you think?

Reply:
Actually on second thought I think I will try to use regular methods as you suggested and if it still works properly I'll stick to that, it would be less confusing for everybody